services/inputhost/exthost.go

```diff
@@ -853,3 +865,49 @@ func (conn *extHost) getState() *admin.InputDestExtent {
 
 	return ext
 }
+
+// stopWritePump is used to stop the write pump if the extent is active and
+// mark the extHost as "Draining"
+func (conn *extHost) stopWritePump() {
+	if conn.state == extHostActive {
+		close(conn.closeChannel)
+		conn.state = extHostDraining
+		// wait for the write pump to drain for some timeout period
+		if ret := common.AwaitWaitGroup(&conn.waitWriteWG, defaultWGTimeout); !ret {
+			conn.logger.Fatalf("unable to stop write pump; wait group timeout")
+		}
+	}
+}
+
+// isDrained returns true if the following conditions are true:
+// 	(1) extHost state is "extHostDraining"
+//	(2) we have seen the reply for the last sent sequence number
+// if not, it returns false
+// Note: RLock() is sufficient here.
+func (conn *extHost) isDrained(replySeqNo int64) bool {
+	conn.lk.RLock()
+	if conn.state != extHostDraining {
```

Comment:
This is mainly to "unblock" the readPump once it has aggregated everything. So I would call it isDrained(). ideally if conn.state is inactive, i don't even want to do anything here because at that point we would have closed everything and there is no point in checking the seq no stuff. In summary this should be checked only if the state is in "Draining".. other stuff can be ignored since we are hard failing anyway.